bfdd/ptm_adapter.c

```diff
@@ -453,15 +482,20 @@ static void bfdd_dest_deregister(struct stream *msg, vrf_id_t vrf_id)
 
 	/* Unregister client peer notification. */
 	pcn = pcn_lookup(pc, bs);
-	pcn_free(pcn);
-	if (bs->refcount ||
-	    CHECK_FLAG(bs->flags, BFD_SESS_FLAG_CONFIG))
+	if (pcn != NULL) {
+		pcn_free(pcn);
 		return;
+	}
 
-	bs->ses_state = PTM_BFD_ADM_DOWN;
-	ptm_bfd_snd(bs, 0);
+	if (bglobal.debug_zebra)
+		zlog_debug("ptm-del-dest: failed to find BFD session");
 
-	ptm_bfd_sess_del(&bpc);
+	/*
+	 * XXX: We either got a double deregistration or the daemon who
+	 * created this is no longer around. Lets try to delete it anyway
+	 * and the worst case is the refcount will detain us.
+	 */
+	_ptm_bfd_session_del(bs, BD_NEIGHBOR_DOWN);
```

Comment:
How do we know if we are removing this due to ZEBRA_BFD_DEST_DEREGISTER message received or client going down? Diag for both the events will be different.

Reply:
This code is exception handling, it should not happen. However we could change `pcn_free` to return different diag/status on shutdown. What is your suggestion?

e.g.
```c
bs->refcount--;
if (bs->refcount > 1) // there are still other users
  _ptm_bfd_session_del(bs, BD_NEIGHBOR_DOWN);
else // we are the last one
  _ptm_bfd_session_del(bs, ???);
```

Reply:
I think it's a good idea to return different diag codes here -- at least then if the error pops up you can do a quick grep to find the error and figure out what's going on. Could help shorten a troubleshooting session in the future a bit.

Reply:
You can figure that out with `debug bfd peer`. I don't oppose changing the diagnostic, but I can't think of a better one to put here.

BTW here are the diagnostics available:

```c
enum bfd_diagnosticis {
	BD_OK = 0,
	/* Control Detection Time Expired. */
	BD_CONTROL_EXPIRED = 1,
	/* Echo Function Failed. */
	BD_ECHO_FAILED = 2,
	/* Neighbor Signaled Session Down. */
	BD_NEIGHBOR_DOWN = 3,
	/* Forwarding Plane Reset. */
	BD_FORWARDING_RESET = 4,
	/* Path Down. */
	BD_PATH_DOWN = 5,
	/* Concatenated Path Down. */
	BD_CONCATPATH_DOWN = 6,
	/* Administratively Down. */
	BD_ADMIN_DOWN = 7,
	/* Reverse Concatenated Path Down. */
	BD_REVCONCATPATH_DOWN = 8,
	/* 9..31: reserved. */
};
```